libp2p/peerstore.nim

```diff
@@ -80,83 +75,82 @@ proc set*[T](peerBook: var PeerBook[T],
-  
-  peerBook.book.mgetOrPut(peerId,
-                          initHashSet[T]()).incl(entry)
-  
-  # Notify clients
-  for handler in peerBook.changeHandlers:
-    handler(peerId, peerBook.get(peerId))
-
-# Helper for seq
-proc set*[T](
-  peerBook: var SetPeerBook[T],
-  peerId: PeerId,
-  entry: seq[T]) =
-  ## Add entry to a given peer. If the peer is not known,
-  ## it will be set with the provided entry.
-  peerBook.set(peerId, entry.toHashSet())
-  
+proc addHandler*[T](peerBook: PeerBook[T], handler: PeerBookChangeHandler) =
+  peerBook.changeHandlers.add(handler)
 
 ##################  
 # Peer Store API #
 ##################
-
-proc addHandlers*(peerStore: PeerStore,
-                  addrChangeHandler: AddrChangeHandler,
-                  protoChangeHandler: ProtoChangeHandler,
-                  keyChangeHandler: KeyChangeHandler) =
-  ## Register event handlers to notify clients of changes in the peer store
-  
-  peerStore.addressBook.changeHandlers.add(addrChangeHandler)
-  peerStore.protoBook.changeHandlers.add(protoChangeHandler)
-  peerStore.keyBook.changeHandlers.add(keyChangeHandler)
-
-proc delete*(peerStore: PeerStore,
-             peerId: PeerId): bool =
+macro getTypeName(t: type): untyped =
+  # Generate unique name in form of Module.Type
+  let typ = getTypeImpl(t)[1]
+  newLit(repr(typ.owner()) & "." & repr(typ))
+
+proc `[]`*[T](p: PeerStore, typ: type[T]): T =
+  let name = getTypeName(T)
+  result = T(p.books.getOrDefault(name))
+  if result.isNil:
+    result = T.new()
+    result.deletor = proc(pid: PeerId) =
+      # Manual method because generic method
+      # don't work
+      discard T(p.books.getOrDefault(name)).del(pid)
+    p.books[name] = result
+  return result
+
+proc del*(peerStore: PeerStore,
+             peerId: PeerId) =
   ## Delete the provided peer from every book.
-  
-  peerStore.addressBook.delete(peerId) and
-  peerStore.protoBook.delete(peerId) and
-  peerStore.keyBook.delete(peerId)
+  for _, book in peerStore.books:
+    book.deletor(peerId)
 
 proc updatePeerInfo*(
   peerStore: PeerStore,
   info: IdentifyInfo) =
 
   if info.addrs.len > 0:
-    peerStore.addressBook.set(info.peerId, info.addrs)
+    peerStore[AddressBook][info.peerId] = info.addrs
 
   if info.agentVersion.isSome:
-    peerStore.agentBook.set(info.peerId, info.agentVersion.get().string)
+    peerStore[AgentBook][info.peerId] = info.agentVersion.get().string
 
   if info.protoVersion.isSome:
-    peerStore.protoVersionBook.set(info.peerId, info.protoVersion.get().string)
+    peerStore[ProtoVersionBook][info.peerId] = info.protoVersion.get().string
 
   if info.protos.len > 0:
-    peerStore.protoBook.set(info.peerId, info.protos)
+    peerStore[ProtoBook][info.peerId] = info.protos
+
+  let cleanupPos = peerStore.toClean.find(info.peerId)
+  if cleanupPos >= 0:
+    peerStore.toClean.delete(cleanupPos)
+
+proc cleanup*(
+  peerStore: PeerStore,
+  peerId: PeerId) =
+
+  if peerStore.capacity <= 0:
```

Comment:
usually `-1` would mean infinite capasity, seems to be reversed here?

Reply:
Added this,
> 0: regular capacity
0: no capacity
< 0: infinite